File: R/ieeeround.R
#
# Set/Get rounding mode functions for R
# 
# Copyright 2010,2011 Gianluca Amato
#

FE.TONEAREST <- 0
FE.DOWNWARD <- 0x400
FE.UPWARD <- 0x800
FE.TOWARDZERO <- 0xc00

fegetround <- function() {
  x = .C("_r_fegetround",rounding.mode = integer(1), PACKAGE="ieeeround")
  return(x$rounding.mode)
}

fesetround <- function(rounding.mode=FE.TONEAREST) {
  x = .C("_r_fesetround", rounding.mode = as.integer(rounding.mode), PACKAGE="ieeeround")
  return(x$rounding.mode)
}

File: src/ieeeround.c
/*
 * Set/Get rounding mode wrapper function for R
 * 
 * Copyright 2010,2011 Gianluca Amato
 */

#include <fenv.h>
#include "ieeeround.h"
#include "R.h"

/* #define DEBUG */

void _r_fegetround(int *rounding_mode) 
{
#ifdef DEBUG
  warning("Starting _r_fegetround. Rounding mode input is %d\n",*rounding_mode);
#endif
  *rounding_mode =  fegetround();
#ifdef DEBUG
  warning("Ending_r_fegetround. Rounding mode input is %d\n",*rounding_mode);
#endif
}

void _r_fesetround(int *rounding_mode)
{
#ifdef DEBUG
  warning("Starting _r_fesetround. Rounding mode input is %d\n",*rounding_mode);
#endif
  *rounding_mode = fesetround(*rounding_mode);
#ifdef DEBUG
  warning("Ending_r_fesetround. Current rounding mode input is %d\n",fegetround());
#endif
}

